### PR TITLE
Implement term.ColorLogger

### PR DESCRIPTION
**Pull request by tgulacsi**
This needs a Hijack method for Logger - this is implemented on Logfmt.
This logger colorizes whole records, for the first matching key/value.

This way you can colorize the record if it contains a given key, or if it contains a non-nil error, for example.
Or colorize levels differently.

### CODE REVIEW

**Comment by ChrisHines** [comment]
There is a lot to review here so I am going to start with a few big things I'd like to see changed, and will hold off on other comments until we see the result of the bigger changes.

First, I think we should change the `NewColorLogger` signature to:
```go
func NewColorLogger(logger log.Logger, color func(keyvals ...interface{}) FgBgColor) log.Logger
```
There are two changes here.

1. The logger parameter is now type `log.Logger`. This change requires using a type assertion in the implementation to get access to the `Hijack` method, but it also keeps variables in client code as type `log.Logger`. Also you can restore the return types of `NewLogfmtLogger` to `log.Logger`.
1. We replace the `opts` variadic parameter with a function that is more flexible and will simplify the implementation of `ColorLogger`. Now `ColorLogger` is doing one thing, coloring the logs; the `color` function decides what the colors to paint; and the wrapped `Logger` controls the textual encoding. Three nicely orthogonal pieces working together.

In concert with the above change we can change `HijackLogger` with
```go
type Hijacker interface {
    Hijack(func(io.Writer) io.Writer)
}
```
I don't think there is any reason for it to embed the `Logger` interface since it is an optional interface. We will require `Logger` in the parameter type, and assert internally that it implements `Hijacker`.

Update the PR if you agree that these are good changes. I think there will be much less to review after that.

File: log/term/colorlogger.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+type FgBgColor struct {
+	Fg, Bg Color
+}
+
+func (c FgBgColor) IsZero() bool {
+	return c.Fg == NoColor && c.Bg == NoColor
+}
+
+type ColorOption struct {
```
Comment (ChrisHines) [inline log/term/colorlogger.go]: ColorOption is not used and should be removed.

File: log/term/colorlogger.go [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	Key   FgBgColor
+	Value func(interface{}) FgBgColor
+}
+
+var _ = log.Logger((*colorLogger)(nil))
+
+// NewColorLogger returns a log.Logger which prouces nice colored logs.
```
Comment (ChrisHines) [inline log/term/colorlogger.go]: s/prouces/produces/

File: log/term/colorlogger.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+// It colors whole records based on the FgBgColor returned by the color function.
+//
+// For example for such a function, see LevelColor.
```
Comment (ChrisHines) [inline log/term/colorlogger.go]: `LevelColor` does not exist.

File: log/term/colorlogger.go [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+		logger: logger,
+	}
+	if hj, ok := logger.(log.Hijacker); ok {
+		cl.color = color // otherwise, no coloring is possible!
+		hj.Hijack(func(w io.Writer) io.Writer {
+			cl.w = w
+			return cl
+		})
+	}
+	return cl
```
Comment (ChrisHines) [inline log/term/colorlogger.go]: Why not return `logger` when we cannot hijack?

File: log/term/colorlogger.go [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+//					return term.FgBgColor{Fg: term.White, Bg: term.Red}
+//				}
+//			}
+//			return term.FgBgColor{}
+//		})
+//
+//	logger.Log("c", "c is uncolored value", "err", nil)
+//	logger.Log("c", "c is colored 'cause err colors it", "err", errors.New("coloring error"))
+func NewColorLogger(logger log.Logger, color func(keyvals ...interface{}) FgBgColor) log.Logger {
+	cl := &colorLogger{
```
Comment (ChrisHines) [inline log/term/colorlogger.go]: Move this inside the following `if` statement.

File: log/term/colorlogger.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	n += m
+	m, err = l.w.Write([]byte("\x1b[0m"))
+	return n + m, err
+}
+
+func (l *colorLogger) Hijack(f func(io.Writer) io.Writer) {
+	l.w = f(l.w)
+}
+
+type ColoredValue struct {
```
Comment (ChrisHines) [inline log/term/colorlogger.go]: `ColoredValue` is not used and can be removed.

File: log/term/colorlogger.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+	if cv.Bg == 0 {
+		return fmt.Sprintf("\x1b[%dm%v\x1b[0m", 30+cv.Fg, cv.Value)
+	}
+	if cv.Fg == 0 {
+		return fmt.Sprintf("\x1b[%dm%v\x1b[0m", 40+cv.Bg, cv.Value)
+	}
+	return fmt.Sprintf("\x1b[%dm\x1b[%dm%v\x1b[0m", 30+cv.Fg, 40+cv.Bg, cv.Value)
+}
+
+func AsString(v interface{}) string {
```
Comment (ChrisHines) [inline log/term/colorlogger.go]: `AsString` is only used in test code, move it to colorlogger_test.go and unexport it.

File: log/term/colorlogger_test.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+	buf.Reset()
+	if err := logger.Log("a", 1, "err", errors.New("error")); err != nil {
+		t.Fatal(err)
+	}
+	if want, have := "\u001b[32m\u001b[48ma=1 err=error\n\u001b[0m", buf.String(); want != have {
+		t.Errorf("want %#v, have %#v", want, have)
+	}
+
+	buf.Reset()
```
Comment (ChrisHines) [inline log/term/colorlogger_test.go]: This section doesn't test anything related to color logging, remove it.

**Comment by ChrisHines** [comment]
Aside from my inline comments above, I'm a little concerned about the concurrency story. We should think about it carefully. I have two concerns.

1. The `Hijacker` interface implementations added to the JSON and Logfmt Loggers introduces potential race conditions to those loggers.
1. The `actColorMu` in `colorLogger` seems awkward and it doesn't prevent potentially strange coloring behavior if the wrapped `Logger` is used directly while concurrently being used by `colorLogger`.

File: log/term/colorlogger_test.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	if err := logger.Log("std_map", map[int]int{1: 2}, "my_map", mymap{0: 0}); err != nil {
+		t.Fatal(err)
+	}
+	if want, have := "std_map=\""+logfmt.ErrUnsupportedValueType.Error()+"\" my_map=special_behavior\n", buf.String(); want != have {
+		t.Errorf("want %#v, have %#v", want, have)
+	}
+}
+
+func newColorLogger(t testing.TB, w io.Writer) log.Logger {
```
Comment (ChrisHines) [inline log/term/colorlogger_test.go]: parameter `t` is not used and can be removed.

**Comment by ChrisHines** [comment]
@tgulacsi Thanks for the updates.

The bad news is that the current code really shows the problems with the `Hijacker` interface. That approach introduces too many edge cases to easily reason about and is too invasive of loggers that we want to wrap with `ColorLogger`.

The good news is that I have come up with an alternate approach that solves all of those problems. I have already coded it, so I'm going to start a new PR building on this one.

I will leave this PR open until we have reviewed the new one and agree it supersedes this one.

**Comment by tgulacsi** [comment]
Ok, I'm curiously waiting for that!

---
https://onename.com/tgulacsi

2015-08-31 6:22 GMT+02:00 Chris Hines <notifications@github.com>:

**Comment by ChrisHines** [comment]
@tgulacsi See PR #116.

**Comment by peterbourgon** [comment]
Can this be closed in deference to #116?

**Comment by tgulacsi** [comment]
Yes.

**Pull request closed without merging**